lib/ClangToAst.ml

```diff
@@ -75,16 +75,22 @@ let add_to_list_lid x data m =
 
 (* ENVIRONMENTS *)
 
+(* For a given tagged union variable, we store the case (variant name) it is currently
+  in, as well as the variable corresponding to the constructor contents pattern *)
+type tagged_case = { case: string; var: string }
+
 type env = {
   (* Variables in the context, with their types, and a reference to tell whether they end up being
-     mutated as some point. *)
-  vars : (string * typ * bool ref) list;
+     mutated as some point.
+    The last reference is used if the variable is a tagged union, to store information about its current state.
+  *)
+  vars : (string * typ * bool ref * tagged_case option ref) list;
```

Comment:
time to introduce a record type?

Reply:
I'm also very weary of having a reference in there -- the environments are persistent, why not have `refine_var_case: env -> env` that gives you a fresh environment for your `then` branch, and the old environment remains a value?

Reply:
I agree, I'm not a fan of the risks involved.
I wanted to avoid yet another traversal of the list, but it's probably not worth it. I'll fix it

Reply:
Record type introduced in [8047139](https://github.com/AeneasVerif/scylla/pull/33/commits/8047139366619950a3d6187837d8896c6d59c56b)

Removing the use of refs was done in [01efdb5](https://github.com/AeneasVerif/scylla/pull/33/commits/01efdb58794722a7e243304ac0265ea1c202ec40)